### PR TITLE
[WIP] systemd.network: module fixes

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change
This PR fixes some wrong assertions currently in the `systemd.network` module, adds support for some not-yet-supported options, and moves some other options around to make everything them more consistent with the rest. 

Contains the DHCP values fix from #35841, but nothing from the controversial "disable IPv6 usecase".

I'd love to get some feedback from people using networkd, in concert with the nixos module

/cc @andir @Mic92 @netixx 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/system/boot/networkd.nix [hunk 1]
```diff
@@ -96,7 +96,7 @@ let
       "Description" "DHCP" "DHCPServer" "IPForward" "IPMasquerade" "IPv4LL" "IPv4LLRoute"
       "LLMNR" "MulticastDNS" "Domains" "Bridge" "Bond" "IPv6PrivacyExtensions"
     ])
-    (assertValueOneOf "DHCP" ["both" "none" "v4" "v6"])
```
Comment (andir) [inline nixos/modules/system/boot/networkd.nix]: Nitpick about the commit message: 
It's that those were wrong. They were simply not being documented anymore. As per systemd policy you do not (immediately) remove those old options but instead drop them from the documentation ;-)

https://github.com/systemd/systemd/blob/master/src/network/networkd-network.c#L832-L840

File: nixos/modules/system/boot/networkd.nix [hunk 2]
```diff
@@ -94,7 +94,7 @@ let
   checkNetwork = checkUnitConfig "Network" [
     (assertOnlyFields [
       "Description" "DHCP" "DHCPServer" "IPForward" "IPMasquerade" "IPv4LL" "IPv4LLRoute"
-      "LLMNR" "MulticastDNS" "Domains" "Bridge" "Bond" "IPv6PrivacyExtensions"
```
Comment (andir) [inline nixos/modules/system/boot/networkd.nix]: So this breaks "old" configurations?

Shouldn't we have a change log entry for that? Or even better support (time limited) backwards compatibility?

File: nixos/modules/system/boot/networkd.nix [hunk 3]
```diff
@@ -94,7 +94,7 @@ let
   checkNetwork = checkUnitConfig "Network" [
     (assertOnlyFields [
       "Description" "DHCP" "DHCPServer" "IPForward" "IPMasquerade" "IPv4LL" "IPv4LLRoute"
-      "LLMNR" "MulticastDNS" "Domains" "Bond" "IPv6PrivacyExtensions"
```
Comment (andir) [inline nixos/modules/system/boot/networkd.nix]: So this breaks "old" configurations?

Shouldn't we have a change log entry for that? Or even better support (time limited) backwards compatibility?

File: nixos/modules/system/boot/networkd.nix [hunk 4]
```diff
@@ -484,6 +484,16 @@ let
       '';
     };
 
+    vrf = mkOption {
```
Comment (andir) [inline nixos/modules/system/boot/networkd.nix]: Did you even test this?

Last time I checked after implementing this in networkd you also had to have a matching `.netdev` file with `Kind=vrf`. Currently we are not allowing `Kind=vrf`. See https://github.com/flokli/nixpkgs/blob/69c091096769ca2ce001beb1072ff0a25812be4e/nixos/modules/system/boot/networkd.nix#L33

**Review by andir — CHANGES_REQUESTED** [review]

File: nixos/modules/system/boot/networkd.nix (hunk 1 above)
Comment (flokli) [reply nixos/modules/system/boot/networkd.nix]: ah, wasn't aware they are still being parsed correctly. Will update the commit message.

File: nixos/modules/system/boot/networkd.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/system/boot/networkd.nix]: yes, it breaks old configurations that uses those attributes before.

As it's currently not the default for network management, plus the module is already pretty huge, I'd be more in favor of breaking/changing this for unstable/18.09 (with a changelog entry), than introducing a compability shim that blows up the code even more.

File: nixos/modules/system/boot/networkd.nix (hunk 3 above)
Comment (flokli) [reply nixos/modules/system/boot/networkd.nix]: see above.

**Comment by globin** [comment]
cc @fpletz

**Comment by netixx** [comment]
Thank you for importing my fixes =)

**Comment by flokli** [comment]
@globin, @fpletz already got a chance to take a look at this?

**Comment by flokli** [comment]
Superseded by #44233 (merged)

**Pull request closed without merging**